src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,238 @@
+import { Address, EstimateGasExecutionError, Hex, encodeFunctionData } from 'viem'
 
-interface ExchangeRate {
+// Apply a multiplier for the decreased swap amount to account for the
+// varying gas costs of different swap providers (or even the same swap)
+const DECREASED_SWAP_AMOUNT_GAS_COST_MULTIPLIER = 1.2
+
+export interface QuoteResult {
   toTokenAddress: string
   fromTokenAddress: string
   swapAmount: BigNumber
   price: string
   provider: string
   estimatedPriceImpact: BigNumber | null
+  preparedTransactions?: PreparedTransactionsResult
+}
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  approveTransaction: TransactionRequestCIP42 & { amountToApprove: bigint }
+  swapTransaction: TransactionRequestCIP42
+}
+
+export interface PreparedTransactionsNeedDecreaseSwapAmountForGas {
+  type: 'need-decrease-swap-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSwapAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSwapAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+function createBaseApproveTransaction(
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction
+) {
+  const { guaranteedPrice, sellTokenAddress, buyAmount, sellAmount, allowanceTarget, from } =
+    unvalidatedSwapTransaction
+  const amountType: string =
+    updatedField === Field.TO ? ('buyAmount' as const) : ('sellAmount' as const)
+
+  const amountToApprove =
+    amountType === 'buyAmount'
+      ? BigInt(new BigNumber(buyAmount).times(guaranteedPrice).toFixed(0, 0))
+      : BigInt(sellAmount)
+
+  const data = encodeFunctionData({
+    abi: erc20.abi,
+    functionName: 'approve',
+    args: [allowanceTarget as Address, amountToApprove],
+  })
+
+  const approveTx: TransactionRequestCIP42 & { amountToApprove: bigint } = {
+    from: from as Address,
+    to: sellTokenAddress as Address,
+    data,
+    type: 'cip42',
+    amountToApprove,
+  }
+
+  return approveTx
+}
+
+function createBaseSwapTransaction(unvalidatedSwapTransaction: SwapTransaction) {
+  const { from, to, value, data, gas } = unvalidatedSwapTransaction
+
+  const swapTx: TransactionRequestCIP42 & { gas: bigint } = {
+    from: from as Address,
+    to: to as Address,
+    value: BigInt(value ? value : 0),
+    data: data as Hex,
+    gas: BigInt(gas), // This isn't entirely accurate
+    type: 'cip42',
+  }
+
+  return swapTx
+}
+
+function getMaxGasCost(txs: TransactionRequestCIP42[]) {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return maxGasCost
+}
+
+async function prepareTransactions(
```

Comment:
wdyt about adding an explicit return type to this function? to make it clearer what comes out, and potentially make type errors easier to debug if we make breaking changes to it by accident